public/client/windows/TracyETW.cpp

```diff
@@ -0,0 +1,248 @@
+
+    ULONG status = StartSession(session);
+    if (status != ERROR_SUCCESS)
+        return {};
+
+    return session;
+}
+
+static ULONG EnableCPUProfiling(Session& session, int microseconds = 125 /* 8KHz = 125us */) {
+    // CPU Profiling requires special privileges on top of admin privileges
+    DWORD access = ElevatePrivilege(SE_SYSTEM_PROFILE_NAME);
+    if (access != ERROR_SUCCESS)
+        return access;
+
+    ULONG status = EnableProvider(session, SystemProfileProviderGuid);
+    if (status != ERROR_SUCCESS)
+        return status;
+
+    TRACE_PROFILE_INTERVAL interval = {};
+    interval.Source = 0; // 0: ProfileTime
+    interval.Interval = (microseconds * 1000) / 100; // in 100's of nanoseconds
+    CONTROLTRACE_ID TraceId = 0; // must be zero for TraceSampledProfileIntervalInfo
+    status = TraceSetInformation(TraceId, TraceSampledProfileIntervalInfo, &interval, sizeof(interval));
+    if (status != ERROR_SUCCESS)
+        return status;
+
+    status = EnableStackWalk(session, PerfInfoGuid, 46);  // PerfInfoGuid Opcode 46: SampledProfile event
+    return status;
+}
+
+static ULONG EnableContextSwitchMonitoring(Session& session) {
+    ULONGLONG MatchAnyKeyword = SYSTEM_SCHEDULER_KW_CONTEXT_SWITCH;
+    ULONG status = EnableProvider(session, SystemSchedulerProviderGuid,
+        EVENT_CONTROL_CODE_ENABLE_PROVIDER, TRACE_LEVEL_INFORMATION, MatchAnyKeyword);
+    if (status != ERROR_SUCCESS)
+        return status;
+    status = EnableStackWalk(session, ThreadGuid, 36);  // ThreadGuid Opcode 36: CSwitch event
+    return status;
+}
+
+static ULONG EnableVSyncMonitoring(Session& session) {
+// TODO: is this correct?
+#if (_WIN32_WINNT < _WIN32_WINNT_WINBLUE) || defined(__MINGW32__)
+    return ETWError(ERROR_NOT_SUPPORTED);
+#endif
+
+    enum Keyword : ULONGLONG {
+        DxgKrnlBase    = 0x0000'0000'0000'0001,    // Microsoft-Windows-DxgKrnl: Base
+        DxgKrnlPresent = 0x0000'0000'0800'0000,    // Microsoft-Windows-DxgKrnl: Present
+        MSFTReserved62 = 0x4000'0000'0000'0000     // winmeta.h: WINEVENT_KEYWORD_RESERVED_62
+    };
+    ULONGLONG MatchAnyKeyword = Keyword::MSFTReserved62 | Keyword::DxgKrnlPresent | Keyword::DxgKrnlBase;
+    ULONGLONG MatchAllKeyword = MatchAnyKeyword;
+
+    EVENT_FILTER_EVENT_ID fe = {};
+    fe.FilterIn = TRUE;
+    fe.Count = 1;
+    fe.Events[0] = 0x0011;  // 0x11 = 17 : VSyncDPC_Info
+    EVENT_FILTER_DESCRIPTOR desc = {};
+    desc.Ptr = (ULONGLONG)&fe;
+    desc.Size = sizeof(fe);
+    desc.Type = EVENT_FILTER_TYPE_EVENT_ID;
+    ENABLE_TRACE_PARAMETERS EnableParamaters = {};
+    EnableParamaters.Version = ENABLE_TRACE_PARAMETERS_VERSION_2;
+    EnableParamaters.EnableProperty = EVENT_ENABLE_PROPERTY_IGNORE_KEYWORD_0;
+    EnableParamaters.SourceId = DxgKrnlGuid;    // or NullGuid? Does it even matter?
+    EnableParamaters.EnableFilterDesc = &desc;
+    EnableParamaters.FilterDescCount = 1;
+
+    ULONG status = EnableProvider(session, DxgKrnlGuid,
```

Comment:
Note for future self: we might need `EVENT_CONTROL_CODE_CAPTURE_STATE` for other events, see https://github.com/google/UIforETW/issues/124#issuecomment-509838160 and https://github.com/GameTechDev/PresentMon/blob/d4385a861a2f4ed2c6db08c875bddf4d6958d797/PresentData/PresentMonTraceSession.cpp#L736